key.go

```
@@ -0,0 +1,806 @@
+		return KeyOpEncrypt
+	case "decrypt":
+		return KeyOpDecrypt
+	case "wrapKey":
+		return KeyOpWrapKey
+	case "unwrapKey":
+		return KeyOpUnwrapKey
+	case "deriveKey":
+		return KeyOpDeriveKey
+	case "deriveBits":
+		return KeyOpDeriveBits
+	default:
+		return KeyOpInvalid
+	}
+}
+
+func (ko KeyOp) String() string {
+	switch ko {
+	case KeyOpSign:
+		return "sign"
+	case KeyOpVerify:
+		return "verify"
+	case KeyOpEncrypt:
+		return "encrypt"
+	case KeyOpDecrypt:
+		return "decrypt"
+	case KeyOpWrapKey:
+		return "wrapKey"
+	case KeyOpUnwrapKey:
+		return "unwrapKey"
+	case KeyOpDeriveKey:
+		return "deriveKey"
+	case KeyOpDeriveBits:
+		return "deriveBits"
+	case KeyOpMACCreate:
+		return "MAC create"
+	case KeyOpMACVerify:
+		return "MAC verify"
+	default:
+		return "unknown key_op value " + strconv.Itoa(int(ko))
+	}
+}
+
+func (ko KeyOp) IsSupported() bool {
+	return ko >= 1 && ko <= 10
+}
+
+// MarshalCBOR marshals the KeyOp as a CBOR int.
+func (ko KeyOp) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int64(ko))
+}
+
+// UnmarshalCBOR populates the KeyOp from the provided CBOR value (must be int
+// or tstr).
+func (ko *KeyOp) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key_ops value %w", err)
+	}
+
+	if raw.IsString() {
+		v := KeyOpFromString(raw.String())
+		if v == KeyOpInvalid {
+			return fmt.Errorf("unknown key_ops value %q", raw.String())
+		}
+
+		*ko = v
+	} else {
+		v := raw.Int()
+		*ko = KeyOp(v)
+
+		if !ko.IsSupported() {
+			return fmt.Errorf("unknown key_ops value %d", v)
+		}
+	}
+
+	return nil
+}
+
+const (
+	// Octet Key Pair
+	KeyTypeOkp KeyType = 1
+	// Elliptic Curve Keys w/ x- and y-coordinate pair
+	KeyTypeEc2 KeyType = 2
+	// Symmetric Keys
+	KeyTypeSymmetric KeyType = 4
+)
+
+type KeyType int64
+
+// KeyTypeFromString returns the KeyType corresponding to the specified name.
+func KeyTypeFromString(v string) (KeyType, error) {
```

If we can define `KeyTypeInvalid` (see previous comment) then we don't need to return an error here.